충남대 FE_강병현_1주차_과제 Step1

kang-kibong opened the pull request:
안녕하세요. 카테캠 프론트엔드 파트 수강 중인 충남대학교 재학중인 강병현입니다. 😊

먼저 코드리뷰에 도움 주셔서 감사드리며, 과제를 다음과 같이 진행해보았습니다.

## 👨🏻‍💻 해결 과정

- **craco 라이브러리**를 활용하여 `src` 폴더와 각 기능별 폴더들의 절대 경로를 aliasing했습니다.
- `tsconfig.json`에 Prettier와 ESLint 실행을 위한 설정을 추가하고, 절대 경로 설정을 `tsconfig.paths.json`으로 분리하여 확장했습니다. 추가로 lint-staged 명령어 정의와, `prepare` 스크립트를 통해 husky를 자동으로 설치 및 실행할 수 있도록 설정했습니다.
- Emotion 스타일 라이브러리를 통해 간단한 Button 컴포넌트를 만들어 styled가 적용되는지 확인했습니다.
- `resetStyles.ts`에서 Emotion의 css를 사용하여 reset CSS를 작성하고, Global 스타일을 적용했습니다.
- 프로젝트 생성 시 발생하는 불필요한 코드와 파일을 정리했습니다.
- 초기 폴더 구조는 각 기능별로 나누어 구성했습니다.
- Storybook 설치와 더불어 절대 경로 이슈를 `tsconfig-paths-webpack-plugin` 설치 및 Storybook config파일 설정을 통해 해결하였습니다.
- `components` 구조에 따라 파일들을 그룹핑하여 관리하고 story와 docs를 함께 작성해보았습니다.

## 🛠️ 코드 리뷰 반영
### 📄 요구 사항
- [x] babel-preset-react-app 설치
- [x] package.json의 eslintConfig 제거
- [x] testing library와 typescript 패키지 devdependencies로 변경
- [x] README에 초기 폴더 구조에 대한 설명 기재
- [x] lint 대상 변경
- [x] QUESTION.md 추가

## 🙋‍♂️ 리뷰어분께

질문 사항이 많아, QUESTION.md에 기재하였습니다.

위 내용을 바탕으로 코드 리뷰 진행해주시면 감사하겠습니다. 🙇‍♂️

kang-kibong committed:
docs(README.md): add a list of features

kang-kibong committed:
feat(kakao-gift): create TypeScript project using CRA

Initialized a new project using Create React App with TypeScript template.

kang-kibong committed:
chore: remove unnecessary code and files

Removed redundant code and deleted unused files to clean up the project.

kang-kibong committed:
chore: configure absolute paths using craco

Set up absolute import paths using craco to simplify module imports and improve project structure.

kang-kibong committed:
chore: set up Prettier formatting rules

Installed Prettier and configured formatting rules across the project.

kang-kibong committed:
chore: configure ESLint rules according

Installed ESLint and set up rules to enforce coding standards and best practices throughout the project.

kang-kibong committed:
chore: add emotion styling library

Installed @emotion/react and @emotion/styled as project dependencies.

kang-kibong committed:
feat: add Button component

kang-kibong committed:
feat: Apply reset CSS globally

Added a global reset CSS file to normalize styles across the application.

kang-kibong committed:
feat: Initialize folder structure

Created initial folder structure for the project.

kang-kibong committed:
chore: Moved files to new directory structure

Organized project files into a new directory structure for better organization and maintainability.

kang-kibong committed:
feat: feat: Add lint-staged and husky

Integrated lint-staged and husky to enforce linting and formatting on staged files before commits.

kang-kibong committed:
docs: Add feature list for Storybook components in README

Added the detailed feature list for implementing reusable components using Storybook in the README file. The feature list includes:

- Installation of Storybook
- Implementation of Button component with theme, size, and basic attributes
- Implementation of Input component with disabled, invalid, size, and basic attributes
- Implementation of Image component with ratio, radius, and basic attributes

kang-kibong committed:
chore: install and configure Storybook

kang-kibong committed:
refactor(components): update Button component

- Add customizable theme and size props
- Update types for ButtonTheme and ButtonSize
- Set default props for theme and size

kang-kibong committed:
design(Button): Implement button styling based on size props

- Implemented conditional styling for button sizes using getSizeStyles function.
- Applied responsive design for 'responsive' size using media queries.

kang-kibong committed:
design(Button): Implement button styling based on theme props

- Implemented conditional styling for button theme using getThemeStyles function.

kang-kibong committed:
feat(storybook): add Button component story

- Added Storybook configuration for the Button component
- Implemented Button component stories with various themes and sizes

taehwanno commented on src/components/Button/Button.styled.ts:
```diff
@@ -0,0 +1,8 @@
+  color: red;
+`;
+
+export default Layout;
```
컴포넌트를 미리 작게 나누기보다는 기능을 구현하면서 분리해야할 명확한 사례가 생길 때 분리하는게 좋습니다. 현재 Layout은 버튼을 의미하는데 일반적으로 Layout 컴포넌트는 페이지의 여러 컴포넌트에 대한 레이아웃을 결정하는 책임으로 분리하는 케이스가 많기도 하고 지금은 버튼 컴포넌트를 의미하는데 실제 컴포넌트 이름은 Layout 이다보니 자주 헷갈릴 거 같아요!

그래서 지금 단계에서는 분리하지 않고 `src/components/Button/Button.tsx` 안에서 선언하셔서 사용하는게 좋겠습니다.

taehwanno commented on src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,20 @@
+      {text}
+    </Layout>
+  );
+}
```
`Button`은 `button`을 확장하는 컴포넌트라서 타입이 button을 대부분 따라가야하기도 하고 [emotion styled component](https://emotion.sh/docs/styled)를 사용하신다면 지금 작성하신 코드는 아래와 동일한 의미를 가집니다. 그래서 Layout과 따로 Button을 정의해주지 않고 아래처럼 변경하는게 좋지 않을까 싶어요!

```tsx
const Button = styled.button`
  border: 1px solid #000;
  color: red;
`;
```

taehwanno commented on package.json:
```diff
@@ -0,0 +1,67 @@
+    "test": "craco test",
+    "eject": "craco eject",
+    "format": "prettier --write --cache src/**/*",
+    "lint": "eslint --cache .",
```
```suggestion
    "lint": "eslint --cache src",
```

lint 실행해보니 에러가 많이 발생하네요. 아마 node_modules에 포함된 코드도 lint 대상으로 잡혀서 그런거 같은데 한번 살펴보시겠어요? 위에 처럼 수정하면 아마 될 거 같은데 체크 부탁드리겠습니다.

taehwanno commented on README.md:
```diff
@@ -1,2 +1,13 @@
+- [x] reset css 적용
+- [x] .gitignore 추가
+- [x] 불필요한 코드 및 파일 정리
+- [x] 폴더 구조 생성
```
PR 본문에 질문 주신건 PR 리뷰 코멘트로 답변을 드리는게 나을 거 같은데 `QUESTIONS.md` 파일을 하나 만들어서 아래처럼 링크 첨부 후 질문 제목만 적으셔서 커밋해주시면 제가 답변하는식으로 할까 하는데 어떠실까요?

```
## 1주차 [질문](https://kibong.notion.site/db60b098e56e4333ae78e8c44edd09f5)
1. reportWebVital?
2. 절대경로 ESLint 에러 이슈 어떻게 해결해야할까요?
3. ...
```

kang-kibong committed:
feat: add Input component

kang-kibong committed:
feat: add disabled prop styling to Input component

kang-kibong committed:
feat: add invalid prop styling to Input component

kang-kibong committed:
feat: implement size prop styling for Input component

kang-kibong committed:
feat: add Storybook stories for Input component

kang-kibong committed:
feat: add Image component

kang-kibong committed:
feat: add ratio prop to set image aspect ratio in Image component

kang-kibong committed:
feat: add radius prop to set border radius in Image component

kang-kibong committed:
feat: add width and height props to Image component

kang-kibong committed:
feat: add Storybook stories for RatioSquare, RadiusCircle, and RadiusRound in Image component

kang-kibong committed:
docs(README): add README documentation

- Implement GoodsItem component with Default and Ranking variations
  - GoodsItem accepts common props: imageSrc, subtitle, title, amount
  - Ranking variation accepts additional rankingIndex prop
  - Display pink badge for rankingIndex 1-3 and gray badge for other numbers
- Implement Grid component
- Implement Container component

kang-kibong committed:
feat: implement GoodsItem component with common props

- Implemented GoodsItem component
- Accepts common props: imageSrc, subtitle, title, amount
- Renders image, subtitle, title, and amount with the provided props

kang-kibong committed:
feat: add GoodsItem component

kang-kibong committed:
feat: add Ranking component

kang-kibong committed:
feat: add Storybook stories for GoodsItem Default and Ranking components

kang-kibong committed:
feat: add Container component

kang-kibong committed:
feat: add Storybook stories for Container component

kang-kibong committed:
feat: add FullScreen story for Container component

kang-kibong committed:
feat: add Grid component

kang-kibong committed:
feat: add NumberColumns story for Grid component

kang-kibong committed:
feat: add ResponsiveColumns story for Grid component

kang-kibong committed:
chore: move Button component file to common folder

kang-kibong committed:
chore: move files

kang-kibong committed:
docs: update README

kang-kibong committed:
feat: add autodocs tags to all Storybook files

kang-kibong committed:
fix: add babel-preset-react-app to package.json

kang-kibong committed:
docs: add requirements from code review to README

kang-kibong committed:
chore: remove eslintConfig from package.json

kang-kibong committed:
chore: move testing and types dependencies to devDependencies

kang-kibong committed:
docs: add initial folder structure explanation to README

kang-kibong committed:
chore: update lint script target

kang-kibong committed:
docs: create QUESTIONS.md

taehwanno commented on package.json:
```diff
@@ -0,0 +1,67 @@
+    "prettier": "^3.3.2",
+    "typescript-eslint": "^7.14.1"
+  }
+}
```
주의 메세지 하단을 읽어보니 babel-preset-react-app은 유지보수 되지 않으니 `@babel/plugin-proposal-private-property-in-object`을 설치하라고 안내되어 있네요. 프로젝트 진행하는데 문제는 없을거라 어느 쪽이든 선택하셔서 진행하면 되겠습니다.

kang-kibong committed:
feat(button): update responsive height styles

kang-kibong committed:
refactor(ranking): remove magic number by defining a constant

Replaced the magic number 3 used for ranking comparison with a constant TOP_RANKING_THRESHOLD to improve code readability and maintainability.

kang-kibong committed:
refactor(props): refactor all files to spread props for cleaner code

kang-kibong committed:
refactor(goods-item): replace magic numbers with constants

Replaced magic numbers for image sizes with constants IMAGE_SIZE_LARGE and IMAGE_SIZE_SMALL to improve code readability and maintainability in the GoodsItem component.

kang-kibong committed:
chore: move styles folder to assets folder

kang-kibong committed:
fix(dependencies): add @babel/plugin-proposal-private-property-in-object to devDependencies

Added @babel/plugin-proposal-private-property-in-object to devDependencies to resolve missing dependency issue in babel-preset-react-app.

kang-kibong committed:
refactor(GoodsItem): separate renderRanking method for better readability

taehwanno commented on QUESTION.md:
```diff
@@ -0,0 +1,112 @@
+초기 React프로젝트를 reportWebVitals로 console로 간단하게 찍어보니 console탭에서 성능에 대한 결과를 객체 형태로 확인할 수 있었습니다.
+결과로 FCP, TTFB, LCP, FID, CLS와 같이 다양한 방면으로 성능 측정이 가능하다는 것을 확인할 수 있었습니다.
+이러한 웹 성능 측정을 회사에서 실제로 많이 고민하고 코드를 짜는지 궁금하고, 한번도 이런 성능 측정을 해본 경험이 없기에 이 부분에 대해서 더 자세히 알려면 어떤식으로, 어떻게 공부하고 경험할 수 있는지 궁금합니다.
+
```
> 이러한 웹 성능 측정을 회사에서 실제로 많이 고민하고 코드를 짜는지 궁금하고

요건 회사에 속한 팀 문화마다 다를 거 같아요. Web Vitals에 있는 수치를 중요시 여긴다면 CI 과정에서 특정 점수 이하로 내려가지 않도록 추적해서 내려간다면 CI를 깨뜨린다거나 PR 병합을 하지 못하게 한다거나 강하게 규칙을 적용할 수도 있을 거 같구요. 제가 속한 조직에서는 강한 규칙보다는 도구를 활용해서 각 수치가 얼만큼 변화하고 있는지 추적해서 대응하는 방향으로 처리하고 있어요. [datadog](https://www.datadoghq.com/dg/real-user-monitoring/client-side-monitoring/) 또는 [sentry](https://sentry.io/for/performance/) 와 같은 도구의 피처가 그 예시입니다.

> 한번도 이런 성능 측정을 해본 경험이 없기에 이 부분에 대해서 더 자세히 알려면 어떤식으로, 어떻게 공부하고 경험할 수 있는지 궁금합니다.

제가 느끼기에는 일반적으로 아래처럼 3가지 단계를 반복적으로 진행하게 되는데요, 이를 위해 수치적 분석을 위한 개념을 이해하고 분석 및 개선 후 수치 변화 추적을 위해 Web Vitals, Lighthouse, Chrome DevTools의 Performance 탭 사용 방식을 익히는 것 정도가 생각나네요. 

1. 문제 식별
2. 진단을 통한 원인 파악 (병목 지점 파악 포함)
3. 개선 후 변화 분석

아래 3가지를 읽고 이해하는게 시작점이 될 수 있지 않나 싶습니다.

1. [Core Web Vitals](https://web.dev/explore/learn-core-web-vitals?hl=ko)
2. [Metrics](https://web.dev/explore/metrics?hl=ko)
3. [Fast load times](https://web.dev/explore/fast?hl=ko)

taehwanno commented on QUESTION.md:
```diff
@@ -0,0 +1,112 @@
+
+구글링해본 결과 일단 `rules`에 `"import/no-unresolved": "off”`하면 해결할 수 있다고 하여 작성해놓은 상태입니다.
+
+이 이슈를 해결할 수 있는 방안은 무엇일까요?
```
2번 문제는 `import/no-unresolved": "off”` 제거하더라도 에러는 발생안하고 있어서 해결된 거 같기도 한데 맞을까요?

taehwanno commented on QUESTION.md:
````diff
@@ -0,0 +1,112 @@
+└── Button.tsx
+```
+
+보통 어떤식으로 하는지 궁금합니다! 정답이란게 있을까요?
````
책임 분리를 통해 얻는 이점이 무엇인지 명확히 할 필요가 있겠습니다. 어떤 이점이 있을까요?

taehwanno commented on QUESTION.md:
```diff
@@ -0,0 +1,112 @@
+
+스타일을 압축 및 호이스팅하여 최적화하고 souce map및 label로 더나은 개발 경험을 생성하는 플러그인을 원하면 권장한다고 합니다.
+
+해당 내용에 대해서 잘 이해가 가지 않아 고민입니다!
```
혹시 https://emotion.sh/docs/@emotion/babel-plugin 내용 읽어보셨을까요?

taehwanno commented on QUESTION.md:
```diff
@@ -0,0 +1,112 @@
+
+개인적으로 tailwindcss는 inline css와 같이 클래스명 내에서 스타일링할 수 있다는 점에서 유지보수하기 힘들 것 같다는 생각에 평소 css-in-js를 많이 사용하는 것 같습니다.
+
+성능 차이에서일까요? 아니면 편의성때문일까요? 두 라이브러리에 대한 차이에 대해서도 궁금합니다.
```
혹시 [Utility-First Fundamentals](https://tailwindcss.com/docs/utility-first) 읽어보셨을까요? 읽어보신 뒤의 병현 님 생각도 궁금합니다.

kang-kibong replied:
tailwindcss와 css-in-js 의 선택은 **프로젝트의 요구사항**과 **팀의 선호도**에 따라 달라질 수 있다는 생각이 들었습니다.

tailwindcss의 장점은 클래스명에 대해 고민할 필요가 없고, 제공하는 유틸리티 클래스를 사용하여 반응형 디자인과 이벤트 처리를 더욱 간결하게 작성할 수 있다는 점인 것 같습니다. 때문에 기존 CSS를 사용하면 길어지는 코드를 획기적으로 줄일 수 있는점과 css-in-js와는 다르게 모든 클래스를 빌드시간에 생성하기 때문에 성능면에서 좋다는 생각이 들었습니다. 그러나 러닝 커브가 존재하기 때문에 팀원 모두가 사용법에 대해 숙지해야 한다는 점이 있는 것 같습니다.

css-in-js, 특히 styled-components는 동적으로 스타일링하기에 최적화되어 있있는 것 같고 낮은 러닝 커브가 장점인 것 같습니다.
그러나 런타임에서 스타일을 생성하기 때문에 성능 저하가 발생할 수 있다는 점이 단점인 것 같습니다.
최근에는 이를 해결하기 위해 zero-runtime css-in-js 라이브러리가 등장하고 있다는 점을 알게되었습니다.

taehwanno commented on QUESTION.md:
```diff
@@ -0,0 +1,112 @@
+
+이번 과제에서도 적용해보았지만, 이를 고려하면 git hook을 통해 포맷팅와 린팅을 자동화 시킬 수 있다는 점이 장점이라 생각되어 다른 프로젝트에서도 평소 사용하고 있는 편입니다.
+
+현업에서도 많이 쓰이는 기술일까요? 아니면 이것도 필요에 따라 설치하는 편인가요?
```
보통 많이 씁니다. 커밋할 때 eslint, prettier 적용하고 CI에서 잘 적용됐는지 eslint, prettier 실행해서 체크하는 과정을 거치게 되는데 적용하지 않으면 CI에서 깨질 수 있으니까요.

taehwanno commented on QUESTION.md:
```diff
@@ -0,0 +1,112 @@
+
+이번 과제에서는 기능에 따라 패턴을 잡아봤습니다.
+
+혹시 정답이란게 있을까요? 혹은 최근 많이 사용되어지고 있는 패턴이 있을까요?
```
아래 글들 읽어보시고 한번 병현님 생각 전달해봐주시겠어요?

1. [파일 구조](https://ko.legacy.reactjs.org/docs/faq-structure.html)
2. [Colocation](https://kentcdodds.com/blog/colocation)

kang-kibong replied:
일단 “**폴더 구조에 대한 정답은 없다.**”인 것 같습니다.

그러나 도움이 되는 방법 중 하나는 **colocation**을 적용하는 것인 것 같습니다.

처음에는 관련된 기능을 하나의 폴더에 모아 작업을 진행하다가, 점차 프로젝트가 커지면 관련된 파일들을 가능한 한 **가까운 곳**에 배치하는 것이 좋다는 것으로 이해했습니다.

이후에는 특정 기능을 수정할 때 수정해야 할 파일이 어디에 있는지 명확하게 예측할 수 있기 때문에 시간이 단축되고 작업 용이성이 향상될 수 있다는 생각이 들었습니다!
